library/std/src/pipe.rs

```diff
@@ -33,6 +93,61 @@ pub struct PipeWriter(pub(crate) AnonPipe);
 
 impl PipeReader {
     /// Create a new [`PipeReader`] instance that shares the same underlying file description.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// #![feature(anonymous_pipe)]
+    /// # #[cfg(miri)] fn main() {}
+    /// # #[cfg(not(miri))]
+    /// # use std::process::Command;
+    /// # use std::io::{Read, Write};
+    /// # use std::fs::{self, File};
+    /// # fn main() -> std::io::Result<()> {
+    /// const NUM_SLOT: u8 = 2;
+    /// const NUM_PROC: u8 = 5;
+    /// const OUTPUT: &str = "output.txt";
+    ///
+    /// let jobserver = [b'|'; NUM_SLOT as usize];
+    /// let mut jobs = vec![];
+    /// let mut file = File::create_new(OUTPUT)?;
+    ///
+    /// let (reader, mut writer) = std::pipe::pipe()?;
+    ///
+    /// for j in 0..NUM_PROC {
+    ///     jobs.push(
+    ///         Command::new("python")
+    ///             .args([
+    ///                 "-c",
+    ///                 &format!(
+    ///                     "from sys import stdout, stdin\n\
+    ///                      stdin.read(1)\n\
+    ///                      with open('{}', 'a') as fp: fp.write('x')\n\
+    ///                      stdout.write(b'|'.decode())",
+    ///                      OUTPUT
+    ///                 ),
+    ///             ])
```

Comment:
I hope I caught your drift on this one. I couldn't preload the pipe and get the same nifty semaphore behavior because (as far as I could figure out) `tee` can't be told to read one byte at a time. So the new example is less interesting, though much simpler.

Reply:
That's true, the only alternative I can think of is

```
Command::new("bash")
    .args([
        "-c",
        format(
            r#"LANG=C IFS= read -r -d '' -n 1 foo
            echo -n 1 >> "{}"
            echo -n 1
            "#,
            OUTPUT,
        )
    )
```

I can't say this is simpler because `read` is actually a feature many using bash doesn't know.

On the other hand it might run on playground.

Reply:
We could use threading instead of processing to make everything pure-rust.

Since it is simulating a jobserver, it'd be fine to use a thread because in reality we do use jobserver within threads as well to synchronise with another process.

Just my two cents.

Reply:
I love your first example, and it's not necessarily a bad thing if people run `help read`, like I did :).